rlbot_gui/bot_management/downloader.py

```diff
@@ -87,24 +112,127 @@ def unzip_callback(self):
+                                 local_subfolder_name=f"{repo_name}-{branch_name}",
                                  clobber=True,
                                  progress_callback=self.zip_download_callback,
                                  unzip_callback=self.unzip_callback)
+        
+        if success:
+            settings = load_settings()
+            settings.setValue(RELEASE_TAG, latest_release["tag_name"])
+
+        return success
+
+
+class BotpackUpdater:
+    """
+    Updates the botpack while updating the progress bar and status text.
+    """
+    def __init__(self):
+        self.status = ''
+
+        self.total_steps = 0
+        self.current_step = 0
+
+    def update_progressbar_and_status(self, status=None):
+        total_progress_percent = int(min(self.current_step / self.total_steps, 1) * 100) if self.total_steps != 0 else 0
+        if status is None: status = self.status
+        status = f'{status} ({total_progress_percent}%)'
+        eel.updateDownloadProgress(total_progress_percent, status)
+
+
+    def download_single(self, tmpdir: Path, repo_url: str, tag: int):
+            download_url = f"{repo_url}/releases/download/incr-{tag}/incremental.zip"
+            downloaded_zip_path = os.path.join(tmpdir, f"downloaded-{tag}.zip")
+            try:
+                urllib.request.urlretrieve(download_url, downloaded_zip_path)
+            except Exception as err:
+                print(err)
+                return False
+            return tag
+
+
+    def update(self, repo_owner: str, repo_name: str, branch_name: str, checkout_folder: Path):
+        repo_full_name = repo_owner + '/' + repo_name
+        repo_url = 'https://github.com/' + repo_full_name
+        master_folder = repo_name + "-" + branch_name
+
+        settings = load_settings()
+        local_release_tag = settings.value(RELEASE_TAG, type=str)
+
+        try:
+            latest_release = get_json_from_url(f"https://api.github.com/repos/{repo_owner}/{repo_name}/releases/latest")
+        except Exception as err:
+            print(err)
+            return False
+
+        if local_release_tag == "" or not os.path.exists(os.path.join(checkout_folder, master_folder)):
+            return "download"
+
+        if local_release_tag == latest_release["tag_name"]:
+            print("The botpack is already up-to-date!")
+            return False
+
+        releases_to_download = list(range(int(local_release_tag.replace("incr-", "")) + 1, int(latest_release["tag_name"].replace("incr-", "")) + 1))
+        if len(releases_to_download) > 50:
+            return "download"
+            
+        local_folder_path = Path(os.path.join(checkout_folder, master_folder))
+
+        self.total_steps = len(releases_to_download)
+        with tempfile.TemporaryDirectory() as tmpdir:
+            with mp.Pool(15) as p:
+                for tag in p.imap(partial(self.download_single, tmpdir, repo_url), releases_to_download):
```

Comment:
I believe this downloads a bunch of incremental releases in parallel, then makes sure they're in the correct order during the for loop. If that's true, please add a code comment explaining.